library/ssl_tls.c

```diff
@@ -1613,6 +1623,23 @@ int mbedtls_ssl_conf_psk( mbedtls_ssl_config *conf,
     if( psk_len > MBEDTLS_PSK_MAX_LEN )
         return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
 
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+    psa_set_key_usage_flags( &key_attributes, PSA_KEY_USAGE_DERIVE );
+    psa_set_key_algorithm( &key_attributes,
+                           PSA_ALG_TLS12_PSK_TO_MS(PSA_ALG_SHA_256) );
+    psa_set_key_enrollment_algorithm( &key_attributes,
```

Comment:
That's OK for TLS 1.2, but TLS 1.3 will need HKDF as well (though PSK support is not implemented yet, work on that should be starting soonish). So perhaps we should use `PSA_ALG_TLS12_PSK_TO_MS(PSA_ALG_ANY_HASH)` here so that we leave `key_enrollment_algorithm` free for TLS 1.3 to use.

Reply:
Actually, question for @gilles-peskine-arm will `PSA_ALG_TLS12_PSK_TO_MS(PSA_ALG_ANY_HASH)` (and later for TLS 1.3 `PSA_ALG_HKDF_EXTRACT(PSA_ALG_ANY_HASH)` or is `PSA_ALG_ANY_HASH` only usable for some algorithms? I just noticed that the documentation doesn't mention `ANY_HASH` as an option [for `PSA_ALG_TLS12_PSK_TO_MS()`](https://armmbed.github.io/mbed-crypto/html/api/ops/kdf.html#c.PSA_ALG_TLS12_PSK_TO_MS) while it does mention it explicitly for [other algs](https://armmbed.github.io/mbed-crypto/html/api/ops/sign.html#c.PSA_ALG_RSA_PKCS1V15_SIGN)

Reply:
[`PSA_ALG_ANY_HASH`](https://armmbed.github.io/mbed-crypto/html/api/ops/sign.html#c.PSA_ALG_ANY_HASH) is only specified and implemented for signature algorithms using the hash-and-sign paradigm. It cannot be used for HMAC, OAEP or key derivation.

Technically it would be easy to generalize. The question is whether it should be.

Reply:
Ok, so for the same of this PK we can keep things are they are, but we'll have to find a solution for 1.3 when we implement it. Could be allowing use of `ANY_HASH` if we think that's the right thing to do, or using a dedicated function for 1.3 PSK provisioning, or an extended function that takes as parameter the version with which the key is meant to be used (I doubt someone would want to use the same PSK for both versions), or we could delay importing the key until the version has been negotiated... We'll see. In the meantime, I withdraw my comment, the code is fine as it is for now.

Reply:
PS: it was right there in the doc, sorry for missing it and thanks for pointing it out.

Reply:
> Technically it would be easy to generalize.

I actually thought of technical difficulties. The first thing that happens to a long HMAC or HKDF key is to hash it. The PSA specification allows implementations to hash at key creation time. Since the key type does not indicate what hash to use, the only way to know is through the policy.

We don't currently do hash during creation and [it's not very likely that we'll do it in the future](https://github.com/Mbed-TLS/mbedtls/pull/5783#discussion_r865191921), but the PSA specification allows it for a tangible benefit (bounding the size of the key, which is useful in implementations without dynamic memory allocation or with limited storage capacity), so we shouldn't extend it in a way that no longer permits it.

Reply:
Good point. But isn't that also an issue for using `psa_set_key_enrollment_algorithm()` on an HMAC or HKDF key? It would mean that an implementation that does key hashing at creation time would need to hash in two ways. Though perhaps that's more acceptable than hashing in as many ways as there are hash algorithms supported in the builds, which can very well be much more than two.

Reply:
`psa_set_key_enrollment_algorithm` is an Mbed TLS extension. It was designed for a specific purpose which doesn't apply to symmetric keys, so we can put limitations to it.

Reply:
In hindsight, it might have been better to embed more information in the key type. For HMAC, distinguishing key types by hash might have made sense. After all you need to know the hash algorithm in order to decide whether two keys are equivalent. But that's minor and definitely not worth changing now.

Reply:
Well, I think `psa_set_key_enrollment_algorithm()` for symmetric keys is coming in quite handy here, as it allows people to provision a PSK that can be used both by SHA-384 and SHA-256 ciphersuites.